hls4ml/model/optimizer/passes/infer_precision.py

```diff
@@ -0,0 +1,290 @@
+import numpy as np
+
+from hls4ml.model.optimizer import OptimizerPass
+from hls4ml.model.types import FixedPrecisionType, UnspecifiedPrecisionType
+
+
+class InferPrecisionTypes(OptimizerPass):
+    def match(self, node):
+        for layer_type in node.types.values():
+            if isinstance(layer_type.precision, UnspecifiedPrecisionType):
+                return True
+        return False
+
+    def transform(self, model, node):
+        types_to_infer = []
+        for type_name, type_obj in node.types.items():
+            if isinstance(type_obj.precision, UnspecifiedPrecisionType):
+                types_to_infer.append(type_name)
+
+        inferred_types = self._infer_precision(node, types_to_infer)
+        for type_name in types_to_infer:
+            if type_name not in inferred_types:
+                self._infer_default_type(node, type_name)
+
+        return False  # No model graph changes made
+
+    def _infer_precision(self, node, types_to_infer):
+        node_class = node.class_name
+        if node_class in ['Dense']:
+            return self._infer_dense_precision(node, types_to_infer)
+
+        if node_class in ['BatchNormalization']:
+            return self._infer_bn_precision(node, types_to_infer)
+
+        if node_class in ['Conv1D', 'Conv2D', 'PointwiseConv1D', 'PointwiseConv2D', 'Conv2DBatchnorm']:
+            return self._infer_conv_precision(node, types_to_infer)
+
+        if node_class in ['SeparableConv1D', 'SeparableConv2D', 'DepthwiseConv2D']:
+            return self._infer_sepconv_precision(node, types_to_infer)
+
+        if node_class in ['Pooling1D', 'Pooling2D']:
+            return self._infer_pooling_precision(node, types_to_infer)
+
+        if node_class in ['Clone', 'Reshape', 'Resize', 'Transpose', 'ZeroPadding1D', 'ZeroPadding2D']:
+            return self._infer_output_matching_precision(node, types_to_infer)
+
+        if node_class in ['Concatenate', 'Merge']:
+            return self._infer_merge_precision(node, types_to_infer)
+
+        # What about quantized activation layer? Setting it to 'auto' manually will break it here. We should prevent
+        # this in config_from_* functions
+
+        return []
+
+    def _infer_default_type(self, node, type_name):
+        model_config = node.model.config
+        default_precision = model_config.backend.convert_precision_string(model_config.model_precision['default'])
+        # No need to change the name of the NamedType since we use the default precision
+        node.types[type_name].precision = default_precision
+
+    def _infer_output_matching_precision(self, node, types_to_infer):
+        assert 'result_t' in types_to_infer and len(types_to_infer) == 1
+
+        in_var = node.get_input_variable()
+        out_var = node.get_output_variable()
+        in_out_type = in_var.type.precision
+        out_var.type.precision = in_out_type
+
+        return ['result_t']
+
+    def _infer_common_precision(self, node, types_to_infer, n_ops):
```

Comment:
I think this assumes integer or fixed precision types. Do we need to handle, e.g., xnor precision types?

Reply:
I should test this PR with the binary model that has Xnor types. I wouldn't expect that you need to infer anything in that case, but perhaps it breaks the optimizer.